source/shared/cpp/ObjectModel/SemanticVersion.h

```diff
@@ -4,55 +4,105 @@
 
 namespace AdaptiveSharedNamespace
 {
-    class SemanticVersion
+    // In a semantic version, this enum tracks which segment we're referring to:
+    // 1.2.3.4
+    // ^ ^ ^ ^- Revision
+    // | | - Build
+    // | - Minor
+    // - Major
+    enum class SemanticVersionSegment : unsigned int
+    {
+        Major = 0,
+        Minor,
+        Build,
+        Revision,
+        Max // must always be last
+    };
+
+    // Indicates if a SemanticVersion instance supports wildcards. We could use a bool here, but it complicates the
+    // SemanticVersion ctors.
+    enum class WildcardSupport
+    {
+        Supported,
+        NotSupported
+    };
+
+    class SemanticVersion final
     {
     public:
-        SemanticVersion(const std::string& version);
+        SemanticVersion(WildcardSupport supportWildcards);
+        SemanticVersion(const std::string& version, WildcardSupport supportWildcards);
 
-        unsigned int GetMajor() const { return _major; }
-        unsigned int GetMinor() const { return _minor; }
-        unsigned int GetBuild() const { return _build; }
-        unsigned int GetRevision() const { return _revision; }
+        // Get the value of the requested version segment
+        unsigned int get(const SemanticVersionSegment segment) const
+        {
+            return _version.at(static_cast<unsigned int>(segment));
+        }
+        bool supportsWildcards() const noexcept { return _supportWildcards == WildcardSupport::Supported; }
+        bool hasWildcard(const SemanticVersionSegment segment) const
+        {
+            return supportsWildcards() && _wildcard.at(static_cast<unsigned int>(segment));
+        }
 
     private:
-        unsigned int _major;
-        unsigned int _minor;
-        unsigned int _build;
-        unsigned int _revision;
+        WildcardSupport _supportWildcards;
+        std::vector<unsigned int> _version;
+        std::vector<bool> _wildcard;
     };
 
     inline bool operator==(const SemanticVersion& lhs, const SemanticVersion& rhs)
     {
-        return lhs.GetMajor() == rhs.GetMajor() && lhs.GetMinor() == rhs.GetMinor() &&
-            lhs.GetBuild() == rhs.GetBuild() && lhs.GetRevision() == rhs.GetRevision();
+        for (unsigned int i = 0; i < static_cast<unsigned int>(SemanticVersionSegment::Max); i++)
+        {
+            const SemanticVersionSegment currSegment = static_cast<SemanticVersionSegment>(i);
+            if (lhs.hasWildcard(currSegment) || rhs.hasWildcard(currSegment))
+            {
+                // wildcards always match.
+                return true;
+            }
+            else if (lhs.get(currSegment) != rhs.get(currSegment))
+            {
+                // not equal, no need to examine further segments.
+                return false;
+            }
+            else
+            {
+                // current segment matches. continue to next.
```

Comment:
Add `continue;`?

Reply:
Sadly, the examples in cppcoreguidelines don't seem to be available. However, according to [ES.77](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-continue), you're supposed to avoid `break;` and `continue;` where possible.

Reply:
Maybe remove the else block then? It looks like it's just asking for someone to add a continue to it :)


---
In reply to: [240842144](https://github.com/Microsoft/AdaptiveCards/pull/2199#discussion_r240842144) [](ancestors = 240842144)